Pet2bids (dicom only)

bendhouseart opened the pull request:
The goals of this PR are very minimal:

- [x] Configure bidscoin such that if the pet2bids plugin is installed use it for PET files instead of anything else
- [x] Convert only dicom PET files at the moment (it's easy enough to update the plugin to target ecat and xlsx after the fact)
- [x] Minimally modify any code outside of the plugin 
- [x] It should be noted that I did start this before bidscoin was using pytest so I need to refactor some of my tests to pytest instead

- More than happy to take up best BIDSCoin practices and modify to suite; at the moment I'm just tickled that this creates a bidsmap in the manner I intended.

- [x] Pass metadata from bidsmap to output from pet2bids (currently it gets lost in translation)
- [x] Update pet2bids to output the whole file name as destination, it's own logic is only extracting the `sub-` and `ses-` name from the target provided by bidscoin.

bendhouseart commented on bidscoin/bids.py:
```diff
@@ -1310,7 +1310,7 @@ def find_run(bidsmap: dict, provenance: str, dataformat: str='', datatype: str='
 
 
-def delete_run(bidsmap: dict, provenance: Union[dict, str], datatype: str= '') -> None:
+def delete_run(bidsmap: dict, provenance: Union[dict, str], datatype: str= '', dataformat: str='') -> None:
```
This is all it took to for me to be able to clean up items that were collected by multiple plugins. Tried to stay away from library code, but couldn't come up with a better way to do this action.

marcelzwiers replied:
That is fine with me, but my first thought was: The user could also avoid this by setting PET as `exclude` datatype in the dcm2niix2bids plugin, and have it converted by your plugin?

marcelzwiers replied:
I mean exclude it in the DICOM tab (which is handled by the dcm2niix2bids plugin), but not in the PET tab

marcelzwiers replied:
Alternatively, if you change the order of the plugins, the dcm2niix4petbids will get executed first, and the dcm2niix2bids plugin will just skip the PET run-item (as it then already exists in the bidsmap)?

marcelzwiers replied:
Excluding is not deleting (in one tab), but that can be a bit confusing for the user indeed. Same holds for the order of the plugins (that can handle the same source data type), but another way out is to remove PET handling altogether from dcm2niix2bids? I mean, it was just a quick and simple hack, better to replace it altogether with a separate plugin?

marcelzwiers replied:
I am happy to limit dcm2niix2bids to MRI only. Have a separate plugin for MRS and another one for PET

bendhouseart replied:
> Excluding is not deleting (in one tab), but that can be a bit confusing for the user indeed. Same holds for the order of the plugins (that can handle the same source data type), but another way out is to remove PET handling altogether from dcm2niix2bids? I mean, it was just a quick and simple hack, better to replace it altogether with a separate plugin?

You've sold me. I just wanted to make sure that we're not losing any functionality, which in theory we shouldn't.

marcelzwiers replied:
I now skip processing DICOM PET data in the dcm2niixplugin (see commit 8a577db2966e564f6a24f252106d35eb9eea8437), So I suppose you can now just delete the PET section in your template bidsmap and use the PET entries in the DICOM section. To do so, in the dcm2niix4pet plugin you should change your dataformat `PET` (which is not a dataformat anyway) to `DICOM`

bendhouseart replied:
> I now skip processing DICOM PET data in the dcm2niixplugin (see commit [8a577db](https://github.com/Donders-Institute/bidscoin/commit/8a577db2966e564f6a24f252106d35eb9eea8437)), So I suppose you can therefore delete the PET section in your template bidsmap and use the PET entries in the DICOM section. To do so, in the dcm2niix4pet plugin you should change your data format `PET` (which is not a data format anyway) to `DICOM`

I was using PET as I figured that `PET` would be better able to capture ECAT, Dicoms w/ modality 'PT', and spreadsheets that contain PET metadata. I haven't integrated handling spreadsheets or ECAT with this PR as I wanted to keep it as small as possible while I got familiar with the codebase. So the question is would you rather have a `PET` data format or add `ECAT` and `Spreadsheet`?  I'm biased towards the former obviously.

bendhouseart commented on bidscoin/heuristics/bidsmap_dccn.yaml:
```diff
@@ -40,6 +40,11 @@ Options:
       meta: [.json, .tsv, .tsv.gz]  # The file extensions of the equally named metadata sourcefiles that are copied over to the BIDS sidecar files
       multiraid: 2                  # The mapVBVD argument for selecting the multiraid Twix file to load (default = 2, i.e. 2nd file)
 #    phys2bidscoin:                  # The settings for the phys2bidscoin plugin
+    dcm2niix4petbids:
+      command: dcm2niix4pet
+      args:
+      anon:
+      meta: [.json, .tsv]
```
I ended up adding the plugin here as it seemed like as good a place as any. I am curious as to whether or not I could adopt the more modular approach that is shown in the example plugin in the read the docs. See

```# The default bids-mappings that are added when installing the plugin
BIDSMAP = {'DemoFormat':{
    'subject': '<<filepath:/sub-(.*?)/>>',     # This filesystem property extracts the subject label from the source directory. NB: Any property or attribute can be used as subject-label, e.g. <PatientID>
    'session': '<<filepath:/ses-(.*?)/>>',     # This filesystem property extracts the session label from the source directory. NB: Any property or attribute can be used as session-label, e.g. <StudyID>

    'func': [                   # ----------------------- All functional runs --------------------
        {'provenance': '',      # The fullpath name of the source file from which the attributes and properties are read. Serves also as a look-up key to find a run in the bidsmap
         'properties':          # The matching (regexp) criteria go in here
            {'filepath': '',    # File folder, e.g. ".*Parkinson.*" or ".*(phantom|bottle).*"
             'filename': '',    # File name, e.g. ".*fmap.*" or ".*(fmap|field.?map|B0.?map).*"
             'filesize': '',    # File size, e.g. "2[4-6]\d MB" for matching files between 240-269 MB
             'nrfiles': ''},    # Number of files in the folder that match the above criteria, e.g. "5/d/d" for matching a number between 500-599
         'attributes':          # The matching (regexp) criteria go in here
            {'ch_num': '.*',
             'filetype': '.*',
             'freq': '.*',
             'ch_name': '.*',
             'units': '.*',
             'trigger_idx': '.*'},
         'bids':
            {'task': '',
             'acq': '',
             'ce': '',
             'dir': '',
             'rec': '',
             'run': '<<1>>',    # This will be updated during bidscoiner runtime (as it depends on the already existing files)
             'recording': '',
             'suffix': 'physio'},
         'meta':                # This is an optional entry for meta-data dictionary that are appended to the json sidecar files
            {'TriggerChannel': '<<trigger_idx>>',
             'ExpectedTimepoints': '<<num_timepoints_found>>',
             'ChannelNames': '<<ch_name>>',
             'Threshold': '<<thr>>',
             'TimeOffset': '<<time_offset>>'}}],

    [...]
```

But I knew I could get things working by adding them to the main template in `bidscoin/heuristics/bidsmap_dccn.yaml`

marcelzwiers replied:
True, I did a first implementation of that some time ago, but I never really tested it very well. The template bidsmap is getting a bit big (with every new plugin), so perhaps it's time to refactor this

marcelzwiers replied:
The advantage of the bidsmap is that every site can have it's own, and it doesn't require modifying the plugin

bendhouseart replied:
Hmm, maybe we should bump all of the plugins out of `bidsmap_dccn.yaml` into each corresponding `plugin.py` file in a separate PR. Then it's not quite so ad hoc.

marcelzwiers replied:
Yes, a separate (dedicated) bidsmap per plugin. It's a bit more complicated to handle but it could be an improvement to the maintenance and modularity on the long run

marcelzwiers replied:
My original thought / idea was that multiple plugin could work using the same mappings (e.g. both dcm2niix2bids and dcm2niix4pet would work with the same DICOM section of the bidsmap), but I guess that doesn't fly very well

marcelzwiers replied:
Separate the logic from the code, it's just that all the logic went into one bidsmap, so that each study could have one bidsmap file with all the info in it to produce the BIDS output

bendhouseart replied:
Like I said, I think my wanting to use that feature could wait until later.

bendhouseart commented on requirements.txt:
```diff
@@ -1,12 +1,13 @@
-pandas
-matplotlib
-numpy
-pydicom>=2
-PyQt5>=5.12.1
-ruamel.yaml>=0.15.35
-coloredlogs
-tqdm>=4.60.0
-multiecho>=0.25
-python-dateutil
-nibabel
-bids_validator
```
I think this is just CRLF to LF, I'm pretty zealous about LF.

bendhouseart commented on tests/test_bids.py:
```diff
@@ -1,116 +1,187 @@
+        anat_provenance = bidsmap['DICOM']['anat'][0]['provenance']
+        func_provenance = bidsmap['DICOM']['func'][0]['provenance']
+
+        # find run with partial provenance
+        not_found_run = bids.find_run(bidsmap=bidsmap, provenance='sub-001', dataformat='DICOM')
+        self.assertFalse(not_found_run)
+
+        # find run with full provenance
+        found_run = bids.find_run(bidsmap=bidsmap, provenance=anat_provenance)
+
+        # create a duplicate provenance but in a different datatype
+        bidsmap['PET'] = bidsmap['DICOM']
+        # mark the entry in the PET section to make sure we're getting the right one
+        tag = 123456789
+        bidsmap['PET']['anat'][0]['properties']['nrfiles'] = tag
+        # locate PET datatype run
+        pet_run = bids.find_run(bidsmap, provenance=anat_provenance, dataformat='PET')
+        self.assertEqual(
+            pet_run['properties']['nrfiles'], tag
+        )
+
+    def test_delete_run(self):
+        # create a copy of the bidsmap
+        with tempfile.TemporaryDirectory() as tempdir:
+            temp_bidsmap = Path(tempdir) / Path(self.full_bidsmap_path.name)
+            shutil.copy(self.full_bidsmap_path, temp_bidsmap)
+            bidsmap, _ = bids.load_bidsmap(temp_bidsmap)
+            anat_provenance = bidsmap['DICOM']['anat'][0]['provenance']
+            # now delete it from the bidsmap
+            bids.delete_run(bidsmap, anat_provenance)
+            self.assertEqual(len(bidsmap['DICOM']['anat']), 0)
+            # verify this gets deleted when rewritten
+            bids.save_bidsmap(_, bidsmap)
+            written_bidsmap, _ = bids.load_bidsmap(_)
+            deleted_run = bids.find_run(written_bidsmap, anat_provenance)
+            self.assertFalse(deleted_run)
+
+
+if __name__ == '__main__':
+    unittest.main()
```
As I mention in the initial PR comment, this was started shortly before pytest was included with bidscoin, this will get refactored to pytest from unittest as the former is just so much better.

marcelzwiers commented:
I can merge if you can fix the test? :-)

bendhouseart commented:
Alright, I managed to get this coining with my lonely test subject 003. I think I'm going to need to do some slight refactoring on PET2BIDS (unless I've just forgotten how to use it) to make sure the filename's passed from BIDSCoin end up in the output. The same with the metadata collected from the bidsmapper, I'm not too surprised that some it isn't getting picked up as we're both shuffling our intermediary files through temporary directories.

However, it should be as easy as passing all of the BIDS metadata collected from the bidsmapper as key pair arguments to the dcm2niix4pet call.

bendhouseart commented:
Added all tests to CI and installed dcm2niix on test runner. I had to do some light refactoring with the installation of spect2nii  and pydeface by adding them to the requirements.txt file. Additionally, added a Makefile so installation and setup is the same on development machines and github's CI. 

Not sure why `pip install -e .[all]` fails on Mac, but I think the changes I added fix it.

Either way, all of your tests now pass, so I can get back to breaking things on my end with the knowledge that the CI and `make test` will now scream at me if I mess up too bad.

bendhouseart commented:
Oh, how do you pick up dcm2niix when it's on windows? Do you just check for it on the Path var?

marcelzwiers commented:
> . The same with the metadata collected from the bidsmapper, I'm not too surprised that some it isn't getting picked up as we're both shuffling our intermediary files through temporary directories.

Do you mean BIDScoin's unpacking of zipped files and/or DICOMDIR data? Otherwise, I think nothing is done in tempdirs?
 
> However, it should be as easy as passing all of the BIDS metadata collected from the bidsmapper as key pair arguments to the dcm2niix4pet call.

BIDScoin has 3 ways of passing metadata:
1) Sidecar files made from the header (such as the JSONs produced by dcm2niix)
2) The sidecar file that is next to the source data (put there by the user beforehand)
3) The metadata entered by the user in the bidseditor

Is (2) something you can work with? I've never used it myself, so you can adapt it (see `bids.copymetadata()`) to your needs (and hence make it a generic mechanism throughout BIDScoin)

marcelzwiers commented:
> I had to do some light refactoring with the installation of spect2nii and pydeface by adding them to the requirements.txt file. Not sure why `pip install -e .[all]` fails on Mac, but I think the changes I added fix it.

Strange, I'll ask my colleague if he can test that too. But if the `requirements.txt` is refactored like this, then we also need to refactor `setup.py` alike (it's a minor detail, I know)

> Additionally, added a Makefile so installation and setup is the same on development machines and github's CI
Great, I am still a CI newbie and had trouble figuring out how to configure github Actions. I tried installing dcm2niix there but failed...

marcelzwiers commented:
> Oh, how do you pick up dcm2niix when it's on windows? Do you just check for it on the Path var?

Well, I don't, I found that to be a wasp nest and I let the user set the path (in the bidsmap, see [here](https://bidscoin.readthedocs.io/en/latest/installation.html#dcm2niix-installation)). In the dccn we use: `command: module add dcm2niix; dcm2niix`

marcelzwiers commented:
p.s. I think `dcm2niix4petbids` is not the best name for the plugin, how about renaming it to just `pet2bids` or `pet2bidscoin`?

bendhouseart commented:
> BIDScoin has 3 ways of passing metadata:
> 1) Sidecar files made from the header (such as the JSONs produced by dcm2niix)
> 2) The sidecar file that is next to the source data (put there by the user beforehand)
> 3) The metadata entered by the user in the bidseditor

> Is (2) something you can work with? I've never used it myself, so you can adapt it (see bids.copymetadata()) to your needs (and hence make it a generic mechanism throughout BIDScoin)

Using 3 was my primary plan and then sending that metadata to the `subprocess.run` call for pet2bids via:
```dcm2niix4pet < input >  -d < output > --kwargs MetadataOne=[0, 1, 2, 3] TimeZero='12:12:12'```

I'll see what all `bids.copy_metadata` does and plan on passing any thing collected from that to the call above via the `--kwargs` flag.

bendhouseart commented:
>Strange, I'll ask my colleague if he can test that too. But if the requirements.txt is refactored like this, then we also need to refactor setup.py alike (it's a minor detail, I know). Or even better, move over to pyproject.toml.

Yeah, I only noticed it on Mac, it behaves fine elsewhere. And a pyproject.toml is the way to go, Poetry too even. It handles packaging so much better than I'm able to with setup.py and all that other wizardry. Everytime I've learned how to package stuff via python I've just as quickly forgotten it... So I just stick with `poetry build` and `poetry publish` for anything I package and send to pypi nowadays.

bendhouseart commented:
>Well, I don't, I found that to be a wasp nest and I let the user set the path (in the bidsmap, see [here](https://bidscoin.readthedocs.io/en/latest/installation.html#dcm2niix-installation)). In the dccn we use: command: module add dcm2niix; dcm2niix

Okay, we do something similar where if we're unable to find it on the system path we ask the user to set it in a config file. I'm sure I'll remember why I asked this once I enable the the CI to run on windows ;)

>p.s. I think dcm2niix4petbids is not the best name for the plugin, how about renaming it to just pet2bids or pet2bidscoin?

You make a very valid point here.

marcelzwiers commented:
> I was referring to this sort of behavior:
> 
> ```
> INFO - dicomsort | >> Sorting: /var/folders/5t/0cjjt4t10hz5_qk4tts9gkv40000gs/T/tmp70tgnxfk/Users/galassiae/Projects/bidscoin/bidscointutorial/raw/sub-003/ses-01 (127 files)
> INFO - dicomsort |    Creating:  /var/folders/5t/0cjjt4t10hz5_qk4tts9gkv40000gs/T/tmp70tgnxfk/Users/galassiae/Projects/bidscoin/bidscointutorial/raw/sub-003/ses-01/30003-Phantom_PetAcquisition_AC Images```
> ```

Oh, but that's because it's tutorial data, in there I did some unsorting for subject 2 as an exercise / demo. But you are right, extended attributes are not working for data that still needs preprocessing before it can be used by BIDScoin directly (this is what dicomsort is doing)

marcelzwiers commented:
> Everytime I've learned how to package stuff via python I've just as quickly forgotten it... So I just stick with `poetry build` and `poetry publish` for anything I package and send to pypi nowadays.
I'd like to move over, it's just that I didn't have to time/energy to port things over from something that I don't fully understand to something else I understand even less :-)

marcelzwiers commented:
Sorry, I noticed that I didn't properly implemented the extended attributes in the DataSource class (I stored it rigidly in self.meta). Will fix it soon

marcelzwiers commented:
I have now updated the DataSource class, the extended attributes are now a class method (i.e. `DataSource._extattributes()`). I'll update the docs as well, but here is already some explanation about how things work.

Adding a json-file to the source data extends the source attributes / gives the exact same results and behaviour as if one would have (over)written that data to (the header of) the source data file itself. So you can do some pre-processing and put json's before running bidscoin -- that data will be transparently available as normal attributes throughout bidscoin.

You can do many nice things now. For instance, you can use these (new or overruling) attributes in a template bidsmap to identify source datatypes (e.g. `{SliceThickness: SliceDim}` if you have `{SliceDim: TheRealSliceThickness}` in your source json). 

Here's another thing you can do. To copy over one or more of the extended attributes to the BIDS sidecar (metadata) files, you can simply put their `<<fieldname(s)>>` in the meta-table -- either in a template bidsmap (then appears in the study bidsmap by default) or in the bidseditor (then the user can/has to add it)! I think this is the route you wanted to take.

However, if you have a long list of extended attributes it can be cumbersome to put them in the metadata table. This is where the `copymetadata()` function comes in (which you were talking about earlier). This is meant to be used by the plugins during bidscoiner runtime, and just transfers **all** extra source metadata to the BIDS folder (for each plugin, the user can control this if needed in the bidseditor by setting the allowed file extensions of the extra source metadata files). A scenario where copymetadata is very useful is e.g. when you have a non-BIDS nifti dataset that was created using dcm2niix and now you want to convert that to BIDS (in that case you just want to copy over all the metadata in the json files). The nibabel2bids plugin can do exactly that. NB: copymetadata is still a very rudimentary function (I never needed it myself so I didn't develop it very much)

Hope this helps :-)

bendhouseart commented:
Sorry for the delay on this and the late reply, I've been completely under the weather for the past 2 weeks or so. 

I'm going to get the CI working as the dbca959 seems to fail there now, but that may just be my own doing merging in your latest changes from master on `Donders-Institute/bidscoin`.

marcelzwiers commented on bidscoin/plugins/pet2bids.py:
```diff
@@ -75,32 +75,36 @@ def is_sourcefile(file: Path) -> str:
 
-    if dataformat == 'PET' and bids.is_dicomfile(sourcefile):
+    for ext in options['meta']:
+        if sourcefile.with_suffix(ext).is_file():
```
I don't know if this would work for you like this? Reading xls metadata can be done from here, but then it is specific for the pet2bids plugin. From BIDScoin's perspective having a source sidecar file with the same name is what I call "extended attributes" and, if you implement your xls reader as such (-> bids.DataSource._extattributes()), then it would be available throughout BIDScoin

bendhouseart replied:
To be honest, I'm not 100% sure on how to implement the spreadsheet portion for PET. But I am sure that I'm going to need an `*.is_spreadsheetfile()` or `*.is_pet_spreadsheetfile()` method to determine if it's PET or not based on what's in the header row.

Then `if pet spreadsheet is True` -> `associate pet spreadsheet w/ imaging data`. It's the later part that's fuzzy on me with respect to BIDSCoin. But, I don't think it's entirely necessary at this point as one can add everything required to make a BIDS Valid PET with the Mapper and Editor as of right now.

marcelzwiers replied:
Mhh, I have no idea what your xls content looks like, can you give me some idea (screenshots or so)? Do you have a different xls for every subject/measurement?

I am happy to give you a hand with getting the xls data into BIDScoin in the best way

marcelzwiers commented:
I made some suggestions and accidentally pushed them :-)

bendhouseart commented:
> I made some suggestions and accidentally pushed them :-)
> 
> What do you think? (see my commit)

I think this works just dandy for Dicoms. I need to double check and see that we're collecting all the required fields in the bids editor, but otherwise it does what I think it should.

Do you think we should add in ECAT's and Spreadsheets or merge this then make smaller PR's for the other formats?

marcelzwiers commented:
I think we can merge it in and do ECAT in a follow-up PR

bendhouseart commented:
Groovy! Thanks!

marcelzwiers commented:
There have been a lot of improvements since v3.7.4, so I am thinking of releasing a new version. Instead of going to v3.8.0, I might jump to v4.0, given that amongst others it is now under CI, there are new bidsapps and there is a new PET plugin :-). I'd like to align it though, with your timeline/plan for getting ECAT support ready. I can wait a bit more, but if it's going to take a while then I may keep that for v4.1?

marcelzwiers commented:
p.s. If you have ideas on architectural changes, this is the time :-)
